src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -1014,6 +1014,31 @@ private ValueTypeShapeCharacteristics ComputeHomogeneousAggregateCharacteristic(
             return NotHA;
         }
 
+        public override bool ComputeIsUnsafeValueType(DefType type)
+        {
+            if (!type.IsValueType)
+                return false;
+
+            MetadataType metadataType = (MetadataType)type;
+            if (metadataType.HasCustomAttribute("System.Runtime.CompilerServices", "UnsafeValueTypeAttribute"))
+                return true;
+
+            foreach (FieldDesc field in metadataType.GetFields())
+            {
+                if (field.IsStatic)
+                    continue;
+
+                TypeDesc fieldType = field.FieldType;
+                if (!fieldType.IsValueType || fieldType.IsPrimitive)
+                    continue;
+
+                if ((DefType)fieldType).IsUnsafeValueType)
```

Comment:
```suggestion
                if (((DefType)fieldType).IsUnsafeValueType)
```

I'm a garbage programmer without IntelliSense. This line needs more LISP.